### DISCUSSION TITLE
Data duplication on server restart

### DISCUSSION THREAD

**Discussion by karthikcodes6** [BitPhinix/slate-yjs#318] [Q&A] [answered]
Data duplication while server restarting.

Steps to reproduce:
1. Type something in the editor.
2. Restart the server now.
3. In the callback onLoadDocument the document comes as empty and again the data get populated to the editor.

```ts
async onLoadDocument(data) {
  if (data.document.isEmpty('content')) {
    const insertDelta = slateNodesToInsertDelta([
      { type: 'p', children: [{ text: 'testing' }] },
    ] as any);
    const sharedRoot = data.document.get('content', Y.XmlText) as Y.XmlText;
    sharedRoot.applyDelta(insertDelta);
  }

  return data.document;
},
```

**Comment by BitPhinix** [comment, marked as answer]
I'm guessing from your previous issue/question you are only persisting the serialized version of the yjs document. You'll have to store the actual yjs document state inside some sort of DB. Hocuspocus offers multiple extensions to do it for you, like `@hocuspocus/extension-redis`:

```ts
import { Server } from '@hocuspocus/server'
import { Redis } from '@hocuspocus/extension-redis'

const server = Server.configure({
  extensions: [
    new Redis({
      // [required] Hostname of your Redis instance
      host: '127.0.0.1',

      // [required] Port of your Redis instance
      port: 6379,
    })
  ],
})

server.listen()
```

**Comment by BitPhinix** [comment]
Sadly hocuspocus is currently still sponsorware (will be fully oss once v1.0 is released), so to get access to the docs you'll have to sponsor @ueberdosis on GitHub I believe.

**Reply by karthikcodes6** [reply]
Thanks for the quick response. We'll become their sponsor. Thanks.

**Comment by Czechh** [comment]
I'm facing this issue if I leave the page editor open for 30mins+. I'm using hocuspocus as well, pretty similar setup as the example.

Is there maybe a strategy missing to handle reconnections after some timeout? 

I was only able to reproduce when I deliberately turn off my wifi as the chrome tools network throttle doesn't work for an open socket.

The socket is good to begin with, I turn off the wifi. The provider starts to retry the connection until I turn on the wifi and it succeds, this then sends back the same doc and the client merges it.

Do you have any suggestions for this problem?

Thank you, for this amazing tool btw.

**Reply by Czechh** [reply]
I'm sorry, after working on this for some time and re-reading what you wrote I understand the difference now between storing the serialized doc and storing the history of changes of the Y doc. I added Redis to help with that and now my test no longer fails!

Again, thank you!

**Comment by cmmartin** [comment]
Hi @BitPhinix 

I have a similar issue when using `slate-yjs` with [y-indexeddb](https://github.com/yjs/y-indexeddb). Each time the slate editor is mounted, the contents are duplicated. The indexeddb updates table also seems to double in size each time the page is loaded. I'm not doing anything special here. Just passing along the y doc which holds my sharedType...

```
 const localDb = React.useMemo(
    () => new IndexeddbPersistence(room, yDoc),
    [room, yDoc]
  )
  ```
  
  Seems like something that should be idempotent is not. Any idea what's going on here?

**Reply by LT1211** [reply]
Hello, have you solved this problem?

**Reply by dpnova** [reply]
@cmmartin are you also using a yjs server? we've noticed some weirdness around this too.

**Comment by DineshCodeFlow** [comment]
The problem is not resolved. Its still happening added `redis` but didn't help
